bayes_kit/rhat.py

```diff
@@ -1,11 +1,88 @@
 import numpy as np
 from numpy.typing import NDArray, ArrayLike
+import scipy as sp
 
 FloatType = np.float64
 VectorType = NDArray[FloatType]
 SeqType = ArrayLike
 
 
+def split_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Return a list of the input chains split in half.  The result will be
+    a list twice as long as the input.  For example,
+    ```
+    >>> split_chains([[1, 2, 3, 4], [5, 6, 7]])
+    [[1, 2], [3, 4], [5, 6], [7]]
+    ```
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of input chains split in half.
+    """
+    return [list(arr) for chain in chains for arr in np.array_split(chain, 2)]
+
+
+def rank_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Returns a copy of the included Markov chains transformed with
+    ranks normalized to [0, 1] and an offset inverse CDF. The ranks
+    are ascending and start with 1 for the smallest value.
+
+    Parameters:
+    chains: list of univariate Markov chains
+
+    Returns:
+    List of chains with values replaced by transformed ranks.
+    """
+    if len(chains) == 0:
+        return chains
+    flattened = np.concatenate(chains)
+    ranks = flattened.argsort().argsort() + 1
+    reshaped_arrays = []
+    current_index = 0
+    for array in chains:
+        size = len(array)
+        reshaped_arrays.append(ranks[current_index : current_index + size])
+        current_index += size
+    return reshaped_arrays
+
+
+def rank_normalize_chains(chains: list[SeqType]) -> list[SeqType]:
+    """Return the rank-normalized version of the input chains.
+
+    The rank-normalized value for element `j` of list `i` is
+    ```
+    inverse_Phi((rank[i][j] - 3/8) / (size(chains) - 1/4),
+    ```
+    where `inv_Phi` is the inverse cumulative distribution function for
+    the standard normal distribution and
+    ```
+    rank[i][j] = rank_chains(chains)[i][j]
+    ```
+    is the rank of element `i` in chain `j` and `size(chains)` is the
+    total number of elements in the chains.
+
+    For a specification of ranking, see :func:`rank_chains`.
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of chains with values replaced by rank-normalized values.
+    """
+    S = sum([len(chain) for chain in chains])
+    result = chains
+    ranked_chains = rank_chains(chains)
+    for i, chain_i in enumerate(ranked_chains):
+        for j, rank_ij in enumerate(chain_i):
+            val = sp.stats.norm.ppf((rank_ij - 0.325) / (S - 0.25))
```

Comment:
I just wanted to note that this is adding a new dependency - previously, scipy was used _only_ in testing

Reply:
Thanks for the reminder.  I updated the README to list the dependency.  

I can't figure out a way around using SciPy---it contains a lot of special functions like inverse CDFs that aren't in NumPy.